modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/report/GridDDMFormFieldTypeReportProcessor.java

```diff
@@ -0,0 +1,137 @@
+import com.liferay.dynamic.data.mapping.model.DDMFormInstanceRecord;
+import com.liferay.dynamic.data.mapping.model.DDMStructure;
+import com.liferay.dynamic.data.mapping.model.Value;
+import com.liferay.dynamic.data.mapping.report.DDMFormFieldTypeReportProcessor;
+import com.liferay.dynamic.data.mapping.service.DDMFormInstanceRecordLocalService;
+import com.liferay.dynamic.data.mapping.storage.DDMFormFieldValue;
+import com.liferay.portal.kernel.json.JSONArray;
+import com.liferay.portal.kernel.json.JSONFactoryUtil;
+import com.liferay.portal.kernel.json.JSONObject;
+import com.liferay.portal.kernel.json.JSONUtil;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Marcos Martins
+ */
+@Component(
+	immediate = true, property = "ddm.form.field.type.name=grid",
+	service = DDMFormFieldTypeReportProcessor.class
+)
+public class GridDDMFormFieldTypeReportProcessor
+	implements DDMFormFieldTypeReportProcessor {
+
+	@Override
+	public JSONObject process(
+			DDMFormFieldValue ddmFormFieldValue, JSONObject fieldJSONObject,
+			long formInstanceRecordId, String ddmFormInstanceReportEvent)
+		throws Exception {
+
+		JSONObject valuesJSONObject = fieldJSONObject.getJSONObject("values");
+
+		Value value = ddmFormFieldValue.getValue();
+
+		JSONObject valueJSONObject = JSONFactoryUtil.createJSONObject(
+			value.getString(value.getDefaultLocale()));
+
+		Iterator<String> iterator = valueJSONObject.keys();
+
+		while (iterator.hasNext()) {
+			String rowName = iterator.next();
+
+			JSONObject rowJSONObject = valuesJSONObject.getJSONObject(rowName);
+
+			if (rowJSONObject == null) {
+				rowJSONObject = JSONFactoryUtil.createJSONObject();
+			}
+
+			String columnName = valueJSONObject.getString(rowName);
+
+			int count = rowJSONObject.getInt(columnName);
+
+			if (ddmFormInstanceReportEvent.equals(
+					DDMFormInstanceReportConstants.EVENT_ADD_RECORD_VERSION)) {
+
+				count++;
+			}
+			else if (ddmFormInstanceReportEvent.equals(
+						DDMFormInstanceReportConstants.
+							EVENT_DELETE_RECORD_VERSION)) {
+
+				count--;
+			}
+
+			rowJSONObject.put(columnName, count);
+
+			valuesJSONObject.put(rowName, rowJSONObject);
+		}
+
+		DDMFormInstanceRecord ddmFormInstanceRecord =
+			ddmFormInstanceRecordLocalService.getDDMFormInstanceRecord(
+				formInstanceRecordId);
+
+		DDMFormInstance ddmFormInstance =
+			ddmFormInstanceRecord.getFormInstance();
+
+		DDMStructure ddmStructure = ddmFormInstance.getStructure();
+
+		DDMFormField ddmFormField = ddmStructure.getDDMFormField(
+			ddmFormFieldValue.getName());
+
+		fieldJSONObject.put(
+			"structure",
+			JSONUtil.put(
+				"columns", _getOptionNames(ddmFormField, "columns")
+			).put(
+				"rows", _getOptionNames(ddmFormField, "rows")
+			));
+
+		return fieldJSONObject;
+	}
+
+	@Reference
+	protected DDMFormInstanceRecordLocalService
```

Comment:
@marcosapmf, does it need to be protected or we can keep it private?

Reply:
@natocesarrego we need it be protected so that it can be mocked on unit tests :)